modules/space/horizonsfile.cpp

```diff
@@ -0,0 +1,877 @@
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A         *
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT    *
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF  *
+ * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE  *
+ * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                                         *
+ ****************************************************************************************/
+
+#include <modules/space/horizonsfile.h>
+
+#include <openspace/util/spicemanager.h>
+#include <openspace/util/time.h>
+#include <ghoul/filesystem/filesystem.h>
+#include <ghoul/fmt.h>
+#include <ghoul/logging/logmanager.h>
+#include <filesystem>
+#include <fstream>
+
+using json = nlohmann::json;
+
+namespace {
+    constexpr std::string_view _loggerCat = "HorizonsFile";
+    constexpr std::string_view ApiSource = "NASA/JPL Horizons API";
+    constexpr std::string_view CurrentVersion = "1.1";
+
+    // Values needed to construct the url for the http request to JPL Horizons API
+    constexpr std::string_view VectorUrl = "https://ssd.jpl.nasa.gov/api/horizons.api?"
+        "format=json&MAKE_EPHEM='YES'&EPHEM_TYPE='VECTORS'&VEC_TABLE='1'&VEC_LABELS='NO'&"
+        "CSV_FORMAT='NO'";
+    constexpr std::string_view ObserverUrl = "https://ssd.jpl.nasa.gov/api/horizons.api?"
+        "format=json&MAKE_EPHEM='YES'&EPHEM_TYPE='OBSERVER'&QUANTITIES='20,33'&"
+        "RANGE_UNITS='KM'&SUPPRESS_RANGE_RATE='YES'&CSV_FORMAT='NO'";
+    constexpr std::string_view Command = "&COMMAND=";
+    constexpr std::string_view Center = "&CENTER=";
+    constexpr std::string_view StartTime = "&START_TIME=";
+    constexpr std::string_view StopTime = "&STOP_TIME=";
+    constexpr std::string_view StepSize = "&STEP_SIZE=";
+
+    // URL encoding
+    constexpr std::string_view WhiteSpace = "%20";
+    constexpr std::string_view HashTag = "%23";
+    constexpr std::string_view DollarSign = "%24";
+    constexpr std::string_view Ampersand = "%26";
+    constexpr std::string_view PlusSign = "%2B";
+    constexpr std::string_view Comma = "%2C";
+    constexpr std::string_view Slash = "%2F";
+    constexpr std::string_view Colon = "%3A";
+    constexpr std::string_view Semicolon = "%3B";
+    constexpr std::string_view EqualsSign = "%3D";
+    constexpr std::string_view QuestionMark = "%3F";
+    constexpr std::string_view AtSymbol = "%40";
+    constexpr std::string_view LeftSquareBracket = "%5B";
+    constexpr std::string_view RightSquareBracket = "%5D";
+
+    std::string replaceAll(std::string string, const std::string& from,
+                           const std::string& to)
+    {
+        if (from.empty())
+            return "";
+
+        size_t startPos = 0;
+        while ((startPos = string.find(from, startPos)) != std::string::npos) {
```

Comment:
The assignment in the while loop looks kind of nasty to me. Makes it a bit difficult to read what's going on (not obvious what is being compared to `std::string::npos`)

Should this be a do-while maybe? Or initialize the variable before the while loop and then later again?